db/db_impl/db_impl.cc

```diff
@@ -4006,19 +4006,19 @@ Status DB::DestroyColumnFamilyHandle(ColumnFamilyHandle* column_family) {
 DB::~DB() {}
 
 Status DBImpl::Close() {
-  if (!closed_) {
-    {
-      InstrumentedMutexLock l(&mutex_);
+  {
+    InstrumentedMutexLock l(&mutex_);
+    if (closed_) {
+      return Status::OK();
+    } else {
       // If there is unreleased snapshot, fail the close call
       if (!snapshots_.empty()) {
         return Status::Aborted("Cannot close DB with unreleased snapshot.");
       }
     }
-
     closed_ = true;
```

Comment:
I think this `closed_` was meant to coordinate with destructor to skip calling `CloseHelper()` a second time. In that case it is reasonable to assume the `Close()` that set `closed_` has already finished.

In this case (concurrent calls to `Close()`) it does not seem to really mean things have been closed. It could still be in progress.

I don't have an exact proposal but wonder if a concurrent `Close()` that returns before things are really closed should instead return `Status::ShutdownInProgress`. Maybe bring in `shutdown_initiated_` to help.

Reply:
Thanks! I probably misread `CloseImpl()` under the mutex lock.

We could add a `closing_mutex_` and wrap the whole call with a lock on that. And in ~DBImpl (for safety in some misuse cases). Then everyone blocks while close is in progress. (I'm suggesting rule: need both DB mutex and closing_mutex_ to write to `closed_` outside of destructor, and thus need either to read `closed_`)

Reply:
Sounds good, making concurrent `Close()`s wait and then return success seems even better.